File: Day_04.m
function [] = Day_04()
%% Load the boards and the numbers
fid = fopen('Day_4_inpt.txt');
tline = fgetl(fid);
line_count = 1;

board_no = 0;
board_line = 0;
tmp_board = zeros(5,5);

while ischar(tline)
    if line_count == 1
        draw_numbers = str2double(split(tline,','));
    elseif isempty(tline)
        if line_count ~= 2
            board(board_no).board = tmp_board;
        end
        board_no = board_no + 1;
        board_line = 1;
    else
        numbers = str2double(split(tline));
        tmp_board(:,board_line) = numbers(~isnan(numbers));
        board_line = board_line +1;
    end
    
    
    tline = fgetl(fid);
    line_count = line_count + 1;
end
board(board_no).board = tmp_board;
fclose(fid);
% Boards loaded

%% Part 1
for board_no = 1:numel(board)
    drawn(board_no).board = zeros(5,5);
end

for number = draw_numbers'
    for board_no = 1:numel(board)
        drawn(board_no).board(board(board_no).board == number) = 1;
    end
    if ~isempty(check_for_win())
        break
    end
end

disp(['Winning value for part 1 is: ' num2str(winvalue(check_for_win()))])

clearvars drawn

%% Part 2
for board_no = 1:numel(board)
    drawn(board_no).board = zeros(5,5);
end

prev_win = [];
for number = draw_numbers'
    for board_no = 1:numel(board)
        drawn(board_no).board(board(board_no).board == number) = 1;
    end
    if numel(check_for_win()) == numel(board)
        break
    end
    prev_win = check_for_win();
end

disp(['Winning value for part 2 is: ' num2str(winvalue(setdiff([1:numel(board)], prev_win)))])

%% Additional functions
    function [winner] = check_for_win()
        winner = [];
        for board_no = 1:numel(board)
            for row = 1:5
                if sum(drawn(board_no).board(row,:)) == 5
                    winner = [winner, board_no];
                end
            end
            for column = 1:5
                if sum(drawn(board_no).board(:,column)) == 5
                    winner = [winner, board_no];
                end
            end
        end
        winner = unique(winner);
    end


    function [value] = winvalue(winner)
        value = sum(sum(board(winner).board(~drawn(winner).board)))*number;
    end

end

File: Day_04.py
import numpy as np

#region: Additional functions
def check_for_win():
    winner = []
    for board_no in range(0,len(boards)):
        for row in range(0,5):
            if np.sum(drawn[board_no][row,:]) == 5:
                winner.append(board_no)
        for column in range(0,5):
            if np.sum(drawn[board_no][:,column]) == 5:
                winner.append(board_no)
    return list(dict.fromkeys(winner))

def bool_equal(array, value):
    tmp_array = np.zeros(np.shape(array), dtype=bool)
    tmp = np.where(array == value)
    for ind in range(0,len(tmp[0])):
        tmp_array[tmp[0][ind], tmp[1][ind]] = True
    return tmp_array


def winvalue(winner):
    return sum(boards[winner][bool_equal(np.array(drawn[winner], dtype=bool), False)])*number
    
#endregion: Addtional functions
#region: Load the bingo boards
file = open('Day_4_inpt.txt', 'r')
lines = file.readlines()

tmp_board = np.empty([5,5])
boards = []
board_no = 0
board_line = 0
for counter,line in enumerate(lines):
    if counter == 0:
        draw_numbers = [int(x) for x in line.split(',')]
    elif line == '\n':
        if counter != 1:
            boards.append(tmp_board.copy())
            board_no += 1
            board_line = 0
    else:
        numbers = [int(x) for x in line.split()]
        tmp_board[board_line, :] = numbers
        board_line += 1
boards.append(tmp_board.copy())
#endregion: Bingo boards loaded
#region: Part 1
drawn = []
for board_no in range(0,len(boards)):
    drawn.append(np.zeros(np.shape(boards[board_no])))

for number in draw_numbers:
    for board_no in range(0,len(boards)):
        drawn[board_no][bool_equal(boards[board_no], number)] = 1
    if len(check_for_win()) > 0:
        break

print('Winning value for part 1 is: ' + str(winvalue(check_for_win()[0])))
#endregion: Part 1 complete!
#region: Part 2
drawn = []
for board_no in range(0,len(boards)):
    drawn.append(np.zeros(np.shape(boards[board_no])))

prev_win = []
for number in draw_numbers:
    for board_no in range(0,len(boards)):
        drawn[board_no][bool_equal(boards[board_no], number)] = 1
    if len(check_for_win()) == len(boards):
        break
    prev_win = check_for_win()

print('Winning value for part 2 is: ' + str(winvalue(list(set(check_for_win()) - set(prev_win))[0])))

#endregion: Part 2 complete!